File: lib/rgfa/field/oriented_identifier_list_gfa2.rb
module RGFA::Field::OrientedIdentifierListGFA2

  def unsafe_decode(string)
    string.split(" ").map do |item|
      OL[item[0..-2].to_sym, item[-1].to_sym]
    end
  end

  def decode(string)
    validate_encoded(string)
    unsafe_decode(string)
  end

  def validate_encoded(string)
    if string !~ /^[!-~][+-]( [!-~][+-])*$/
      raise RGFA::FormatError,
        "#{string.inspect} is not a valid list of oriented GFA2 identifier\n"+
        "(it contains non-printable characters or invalid orientations)"
    end
  end

  def validate_decoded(object)
    case object
    when Array
      object.each do |elem|
        if !elem.kind_of?(RGFA::OrientedLine)
          raise RGFA::TypeError,
            "the array contains an object of class #{elem.class}\n"+
            "(accepted classes: RGFA::OrientedLine)"
        end
        if elem.name !~ /^[!-~]+$/
          raise RGFA::FormatError,
          "the list contains an invalid GFA2 identifier (#{elem.name})\n"+
          "(it contains spaces and/or non-printable characters)"
        end
        if ![:+,:-].include?(elem.orient)
          raise RGFA::ValueError,
            "#{elem} is not a valid GFA2 oriented identifier\n"+
            "(orientation #{elem.orient} is invalid)"
        end
      end
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Array)"
    end
  end

  def unsafe_encode(object)
    case object
    when Array
      object.map do |elem|
        case elem
        when RGFA::OrientedLine
          elem.to_s
        else
          raise RGFA::TypeError,
            "the array contains an object of class #{elem.class}\n"+
            "(accepted classes: RGFA::OrientedLine)"
        end
      end.join(" ")
    when String
      return object
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Array, String)"
    end
  end

  def encode(object)
    validate_decoded(object)
    return unsafe_encode(object)
  end

  module_function :decode
  module_function :unsafe_decode
  module_function :validate_encoded
  module_function :validate_decoded
  module_function :unsafe_encode
  module_function :encode

end

File: gfapy/field/oriented_identifier_list_gfa2.py
import gfapy
import re

def unsafe_decode(string):
  return [ gfapy.OrientedLine(str(l[0:-1]), str(l[-1]))
           for l in string.split(" ")]

def decode(string):
  validate_encoded(string)
  return unsafe_decode(string)

def validate_encoded(string):
  if not re.match(r"^[!-~]+[+-]( [!-~]+[+-])*$", string):
    raise gfapy.FormatError(
      "{} is not a valid list of GFA2 segment names ".format(repr(string))+
      "and orientations")

def validate_decoded(iterable):
  for elem in iterable:
    if not isinstance(elem, gfapy.OrientedLine):
      raise gfapy.TypeError(
            "the list contains an object of class {}\n".format(type(elem))+
            "(accepted classes: gfapy.OrientedLine)")
    elem.validate()
    if not re.match(r"^[!-~]+$", elem.name):
      raise gfapy.FormatError(
        "the list contains an invalid GFA2 identifier {}\n".format(elem.name)+
        "(it contains spaces and/or non-printable characters)")
    if not elem.orient in ["+", "-"]:
      raise gfapy.FormatError(
        "{} is not a valid orientation".format(elem.orient))

def unsafe_encode(obj):
  if isinstance(obj, str):
    return obj
  elif isinstance(obj, list):
    retval = []
    for elem in obj:
      if not isinstance(elem, gfapy.OrientedLine):
        raise gfapy.TypeError(
              "the list contains an object of class {}\n".format(type(elem))+
              "(accepted classes: gfapy.OrientedLine)")
      retval.append(str(elem))
    return " ".join(retval)
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, list)")

def encode(obj):
  validate_decoded(obj)
  return unsafe_encode(obj)
